src/drivers/cmakeDriver.ts

```diff
@@ -514,14 +514,16 @@ export abstract class CMakeDriver implements vscode.Disposable {
 
         log.info(localize('switching.to.kit', 'Switching to kit: {0}', kit.name));
 
-        const opts = this.expansionOptions;
-        opts.vars.buildKit = kit.name;
-        const newBinaryDir = util.lightNormalizePath(await expand.expandString(this.config.buildDirectory, opts));
-
-        const needs_clean = this.binaryDir === newBinaryDir && kitChangeNeedsClean(kit, this._kit);
-        await this.doSetKit(needs_clean, async () => {
+        const oldBinaryDir = this.binaryDir;
+        const needsCleanIfKitChange = kitChangeNeedsClean(kit, this._kit);
+        await this.doSetKit(async () => {
             await this._setKit(kit, preferredGenerators);
             await this._refreshExpansions();
+            const newBinaryDir = util.lightNormalizePath(await expand.expandString(this.config.buildDirectory, this.expansionOptions));
+            const needsCleanIfBuildDirectoryChange = (newBinaryDir !== oldBinaryDir);
+            if (needsCleanIfKitChange || needsCleanIfBuildDirectoryChange) {
```

Comment:
I don't think that changing the build directory is the right condition to delete the cache.  What I expected to see here was "if the kit changed, but the build directory didn't, then remove the prior cache" because I think it makes sense to expect that a different Kit is a different compiler and would configure the project differently (basically the same check that was happening on deleted line 521).  The rest of the fix seems correct because it defers the previous check until after the kit has been set and the `newBinaryDir` can be computed correctly.

**TL;DR:**
I think line 524 should be:
```ts
  if (needsCleanIfKitChange && newBinaryDir === oldBinaryDir) {
```

Reply:
sorry about that. that was the old logic. I changed it in the first commit and I don't remember why. I think I was just confused.

Reply:
just pushed a commit for this.

Reply:
No need to apologize. We appreciate you addressing this issue!